rdtools/soiling.py

```diff
@@ -1052,3 +1065,1530 @@ def monthly_soiling_rates(soiling_interval_summary, min_interval_length=14,
+            max_diff_day = \
+                metric.loc[start_true_vals-f:end_true_vals+f].idxmax()
+            # Set all days in this period as false
+            collapsed_ce.loc[start_true_vals-f:end_true_vals+f] = False
+            collapsed_ce_dummy.loc[start_true_vals-f:end_true_vals+f] = False
+            # Set the max probability day as True (cleaning event)
+            collapsed_ce.loc[max_diff_day] = True
+            # Find the next island of true values
+            start_true_vals = collapsed_ce_dummy.idxmax()
+            if start_true_vals == f:
+                start_true_vals = 0  # Stop iterations
+        else:
+            start_true_vals = 0  # Stop iterations
+    # Return the series of collapsed cleaning events with the original index
+    return pd.Series(index=saveindex, data=collapsed_ce.values)
+
+
+def _rolling_median_ce_detection(x, y, ffill=True, rolling_window=9, tuner=1.5):
+    ''' Finds cleaning events in a time series of performance index (y) '''
+    y = pd.Series(index=x, data=y)
+    if ffill:  # forward fill NaNs in y before running mean
+        rm = y.ffill().rolling(rolling_window, center=True).median()
+    else:  # ... or backfill instead
+        rm = y.bfill().rolling(rolling_window, center=True).median()
+    Q3 = rm.diff().abs().quantile(.75)
+    Q1 = rm.diff().abs().quantile(.25)
+    limit = Q3 + tuner * (Q3 - Q1)
+    cleaning_events = rm.diff() > limit
+    return cleaning_events, rm
+
+
+def _soiling_event_detection(x, y, ffill=True, tuner=5):
+    ''' Finds cleaning events in a time series of performance index (y) '''
+    y = pd.Series(index=x, data=y)
+    if ffill:  # forward fill NaNs in y before running mean
+        rm = y.ffill().rolling(9, center=True).median()
+    else:  # ... or backfill instead
+        rm = y.bfill().rolling(9, center=True).median()
+    Q3 = rm.diff().abs().quantile(.99)
+    Q1 = rm.diff().abs().quantile(.01)
+    limit = Q1 - tuner * (Q3 - Q1)
+    soiling_events = rm.diff() < limit
+    return soiling_events
+
+
+def _make_seasonal_samples(list_of_SCs, sample_nr=10, min_multiplier=0.5,
+                           max_multiplier=2, max_shift=20):
+    ''' Generate seasonal samples by perturbing the amplitude and the phase of
+        a seasonal components found with the fitted CODS model '''
+    samples = pd.DataFrame(index=list_of_SCs[0].index,
+                           columns=range(int(sample_nr*len(list_of_SCs))))
+    # From each fitted signal, we will generate new seaonal components
+    for i, signal in enumerate(list_of_SCs):
+        # Remove beginning and end of signal
+        signal_mean = signal.mean()
+        unique_years = signal.index.year.unique()  # Unique years
+        # Make a signal matrix where each column is a year and each row a date
+        year_matrix = pd.concat([pd.Series(signal.loc[str(year)].values)
+                                 for year in unique_years],
+                                axis=1, ignore_index=True)
```

Comment:
I think there is a bug here when the input series is not a sequence of complete calendar years.  For example if the first year starts on Dec 1, `year_matrix` will align the first year's December with the second year's January.  Here's a quick demonstration and a possible alternative:

<details>
  <summary>Click to expand!</summary>

```python
In [1]: import pandas as pd
   ...: signal = pd.Series(0, index=pd.date_range('2019-06-01', freq='d', periods=1000))
   ...: unique_years = signal.index.year.unique()
   ...: year_matrix = pd.concat([pd.Series(signal.loc[str(year)].values)
   ...:                          for year in unique_years],
   ...:                         axis=1, ignore_index=True)

In [2]: year_matrix
Out[2]: 
       0  1    2    3
0    0.0  0  0.0  0.0
1    0.0  0  0.0  0.0
2    0.0  0  0.0  0.0
3    0.0  0  0.0  0.0
4    0.0  0  0.0  0.0
..   ... ..  ...  ...
361  NaN  0  0.0  NaN
362  NaN  0  0.0  NaN
363  NaN  0  0.0  NaN
364  NaN  0  0.0  NaN
365  NaN  0  NaN  NaN

[366 rows x 4 columns]

In [3]: signal.rename('values').to_frame().assign(
   ...:     doy=signal.index.dayofyear,
   ...:     year=signal.index.year
   ...: ).pivot(index='doy', columns='year', values='values')
Out[3]: 
year  2019  2020  2021  2022
doy                         
1      NaN   0.0   0.0   0.0
2      NaN   0.0   0.0   0.0
3      NaN   0.0   0.0   0.0
4      NaN   0.0   0.0   0.0
5      NaN   0.0   0.0   0.0
..     ...   ...   ...   ...
362    0.0   0.0   0.0   NaN
363    0.0   0.0   0.0   NaN
364    0.0   0.0   0.0   NaN
365    0.0   0.0   0.0   NaN
366    NaN   0.0   NaN   NaN

[366 rows x 4 columns]
```
</details>

Reply:
Thanks for catching this @kanderso-nrel let's make sure we do something sensible about leap years.